lib/tracing-limit/src/lib.rs

```diff
@@ -10,165 +10,200 @@ use std::{
-                            let count = state.count.load(Ordering::Relaxed);
-                            if count > 1 {
-                                let message = format!(
-                                    "Internal log [{}] has been rate limited {} times.",
-                                    state.message,
-                                    count - 1
-                                );
-
-                                self.create_event(&id, &ctx, message, state.limit);
-                            }
-                        }
-                    }
+    // keep track of any span fields we use for grouping rate limiting
+    fn on_record(&self, id: &span::Id, values: &span::Record<'_>, ctx: Context<'_, S>) {
+        {
+            let span = ctx.span(id).expect("Span not found, this is a bug");
+            let mut extensions = span.extensions_mut();
 
-                    return false;
+            match extensions.get_mut::<RateLimitedSpanKeys>() {
+                Some(fields) => {
+                    values.record(fields);
                 }
-            }
+                None => {
+                    let mut fields = RateLimitedSpanKeys::default();
+                    values.record(&mut fields);
+                    extensions.insert(fields);
+                }
+            };
         }
+        self.inner.on_record(id, values, ctx);
+    }
 
-        true
+    #[inline]
+    fn on_follows_from(&self, span: &span::Id, follows: &span::Id, ctx: Context<'_, S>) {
+        self.inner.on_follows_from(span, follows, ctx);
     }
 
-    fn on_event(&self, event: &Event, _ctx: Context<S>) {
-        if is_limited(event.metadata()) {
-            let mut limit_visitor = LimitVisitor::default();
-            event.record(&mut limit_visitor);
+    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
+        let metadata = event.metadata();
 
-            if let Some(limit) = limit_visitor.limit {
-                let start = Instant::now();
+        // if the event is not rate limited, just pass through
+        if !is_limited(metadata) {
+            return self.inner.on_event(event, ctx);
+        }
 
-                let id = event.metadata().callsite();
+        let mut limit_visitor = LimitVisitor::default();
+        event.record(&mut limit_visitor);
+
+        let component_name = {
+            let mut scope = ctx
+                .lookup_current()
+                .into_iter()
+                .flat_map(|span| span.from_root().chain(std::iter::once(span)));
+
+            scope.find_map(|span| {
+                let extensions = span.extensions();
+                extensions
+                    .get::<RateLimitedSpanKeys>()
+                    .and_then(|fields| fields.component_name.clone())
+            })
+        };
+
+        let id = RateKeyIdentifier {
+            callsite: metadata.callsite(),
+            rate_limit_key: limit_visitor.key.clone(),
```

Comment:
I suspect this code is on a hot path. I wonder if it would be beneficial to avoid these clones here by making `RateKeyIdentifier` something like:

```rust
struct RateKeyIdentifier<S: AsRef<str>> {
    callsite: Identifier,
    rate_limit_key: Option<S>,
    component_name: Option<S>,
}
```

Then we would only need to clone the strings when storing the Identifiers.

Reply:
👍 I was thinking about that too. I'll give it a shot.

Reply:
This clone didn't appear to make much of an impact on the benches (see https://github.com/timberio/vector/pull/6563#issuecomment-789205204) so I just left it. I welcome thoughts though!